### ISSUE TITLE
Release 9.2

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#3236]
Quite a few fixes are just about to land.

It'd be great to see the following get in the next release too:

- ~https://github.com/stylelint/stylelint/pull/3195~
- ~https://github.com/stylelint/stylelint/pull/3194~
- ~https://github.com/stylelint/stylelint/pull/3203~
- ~https://github.com/stylelint/stylelint/pull/3176~
- ~https://github.com/stylelint/stylelint/pull/2903~

Any chance of reviews, tweaks on those if we have time?

- [x] stylelint release
- [x] stylelint-config-recommended update/release
- [x] stylelint-config-standard update/release
- [x] stylelint-demo update
- [x] stylelint.io update
- [x] tweet

cc @stylelint/core

**Comment by jeddy3** [comment]
@gucong3000 I'm getting a failed test when trying to do a release on my local machine:

<img width="757" alt="screen shot 2018-03-25 at 23 38 34" src="https://user-images.githubusercontent.com/808227/37880819-22481e22-3086-11e8-96d5-009c9a9d36c0.png">

Am I right in thinking that this is related to this recently merged PR?: https://github.com/stylelint/stylelint/pull/2903

Is there anything we can do to either speed up this test or workaround the timeout period?

**Comment by ntwb** [comment]
I'm also seeing a few slow tests when running `npm test`:

```shell
 PASS  lib/rules/color-hex-length/__tests__/index.js (5.95s)
 PASS  lib/rules/no-duplicate-selectors/__tests__/index.js (6.861s)
 PASS  lib/rules/declaration-property-unit-whitelist/__tests__/index.js (5.304s)
 PASS  lib/rules/at-rule-semicolon-space-before/__tests__/index.js (5.289s)
 PASS  system-tests/001/001.test.js (5.602s)
```

The test `FAIL` you posted @jeddy3 worked for me in that above test run:
```shell
 PASS  system-tests/cli/cli.test.js
```

**Comment by jeddy3** [comment]
Timing out tests in https://github.com/stylelint/stylelint/pull/3241 too.

@stylelint/core Any ideas what might have changed recently to have caused this?

**Comment by gucong3000** [comment]
@jeddy3
> I'm getting a failed test when trying to do a release on my local machine:

Could you try to modify this code, remove `if (parseInt(process.versions.node) < 7)`
https://github.com/stylelint/stylelint/blob/d5d8f77b2408a1a174791a6c55eaacf9aecdd99a/system-tests/cli/cli.test.js#L24-L29

**Comment by jeddy3** [comment]
That seems to have fixed my local copy. But I'm unsure whether it was that change that fixed it or it was because the `cli.test.js` was _the first test to run_ this time.

The tests still feel slow, though. Does anyone know if a change was made, either in stylelint's codebase or in one of our dependencies, that could be causing this?

<img width="599" alt="screen shot 2018-03-26 at 10 01 41" src="https://user-images.githubusercontent.com/808227/37896982-00fce548-30dd-11e8-9257-b062c07b3a22.png">

**Comment by m-allanson** [comment]
I believe Jest caches the time taken for each test to run, then attempts to
run the slowest tests first. I'm not sure if that makes any difference to
timeouts. Check out
https://stackoverflow.com/questions/44866994/how-to-clear-jest-cache.

**Comment by jeddy3** [comment]
> I believe Jest caches the time taken for each test to run, then attempts to
run the slowest tests first.

That would explain the reordering and why builds are passing after restarts.

After running `--clearCache` my local build fails again (even with @gucong3000 change). It feels like we need to make the running of the tests more robust. Any ideas?

<img width="751" alt="screen shot 2018-03-26 at 10 24 45" src="https://user-images.githubusercontent.com/808227/37898145-f05ac590-30df-11e8-9820-10412d79c7bf.png">

Also, do we think this should postpone the release of 9.2?

**Comment by hudochenkov** [comment]
Temporary solution: increase timeout.

`npm run jest:detectleaks` might give us a lead to cause of slowdown.

**Comment by gucong3000** [comment]
```shell
E:\work\gulp-reporter (ansi-escapes -> origin) (gulp-reporter@2.7.4)
$ sh
sh-4.4$ time stylelint --version
8.4.0

real    1m24.783s
user    0m0.075s
sys     0m0.200s
sh-4.4$
```

Run `stylelint --version` on my local machine for more than a minute. (v8.4.0)

**Comment by jeddy3** [comment]
> Run `stylelint --version` on my local machine for more than a minute. **(v8.4.0)**

Ok, since it's been running slow since v8.4 I think we can release v9.2. If not one beats me to it, I can release this evening.

> Temporary solution: increase timeout.

If necessary I'll bump the timeout to get the release out.

---

If anyone has time, I suggest we prioritise https://github.com/stylelint/stylelint/issues/2454 and look to ways we can improve the load time of stylelint (both when bundled for distribution and when running tests). Does anyone have an experience in this area?

**Comment by hudochenkov** [comment]
For me `stylelint --version` is fast ¯﻿\﻿_﻿(﻿ツ﻿)﻿_﻿/﻿¯

```
$ time npx stylelint --version
9.1.3

real	0m1.006s
user	0m0.951s
sys	0m0.126s
```

**Comment by daKmoR** [comment]
for me there is a huge difference...

```
$ time stylelint --version
8.2.0

real    0m0.795s
user    0m0.818s
sys     0m0.044s
```
```
$ time ./node_modules/.bin/stylelint --version
9.1.3

real    0m10.467s
user    0m1.205s
sys     0m2.106s
```

I'm not sure how much difference there could be for globally and locally installed modules...

**Comment by jeddy3** [comment]
Thanks for the data points. It looks like that if any slow down happened, it was somewhere between 8.2.0 and 9.1.3. So, I think we're fine to get 9.2 out and then investigate the speed issue afterwards.

**Comment by jeddy3** [comment]
Unfortunately, even with an increased timeout the tests fail when attempting a release.

However, I believe I'm making progress improving the time taken to run the tests. I've run out of time tonight, though. Hopefully, I'll have time to pick it up tomorrow.

**Comment by ntwb** [comment]
Using Node.js 8.11.0 and npm 5.8.0 I've no issue with v8.0.0 through 9.1.0 /shrug 

```shell
~/dev/stylelint/stylelint master* 
❯ npm i stylelint@8.0.0 -g ;; time -p stylelint --version
  npm i stylelint@8.1.0 -g ;; time -p stylelint --version
  npm i stylelint@8.2.0 -g ;; time -p stylelint --version
  npm i stylelint@8.3.0 -g ;; time -p stylelint --version
  npm i stylelint@8.4.0 -g ;; time -p stylelint --version
  npm i stylelint@9.0.0 -g ;; time -p stylelint --version
  npm i stylelint@9.1.0 -g ;; time -p stylelint --version
```

The timing results with npm install details removed for readability:
```shell
+ stylelint@8.0.0
8.0.0
real         0.65
user         0.62
sys          0.09

+ stylelint@8.1.0
8.1.0
real         0.65
user         0.62
sys          0.08

+ stylelint@8.2.0
8.2.0
real         0.65
user         0.62
sys          0.08

+ stylelint@8.3.0
8.3.0
real         0.85
user         0.80
sys          0.10

+ stylelint@8.4.0
8.4.0
real         0.74
user         0.70
sys          0.10

+ stylelint@9.0.0
9.0.0
real         0.75
user         0.71
sys          0.09

+ stylelint@9.1.0
9.1.0
real         0.77
user         0.70
sys          0.10
```

**Comment by gucong3000** [comment]
> Using Node.js 8.11.0 and npm 5.8.0 I've no issue with v8.0.0 through 9.1.0 /shrug

Your hard drive is SSD?

**Comment by jeddy3** [comment]
Close, but no cigar... I'm still getting a failed test for `"basic"` in `cli.test.js` when attempting a release. (The tests run quicker, though!) 

At first, I thought it was because we're not resolving the promise:

https://github.com/stylelint/stylelint/blob/89d7e5a266b1a0298a70e7c54dc5e45c3b15c8e6/system-tests/cli/cli.test.js#L33-L40

But making that change:

https://github.com/stylelint/stylelint/blob/67b062a24e8f0fe52058ffc8e07566d143a9c7e3/system-tests/cli/cli.test.js#L32-L40

Still results in:

<img width="731" alt="screen shot 2018-03-30 at 10 33 08" src="https://user-images.githubusercontent.com/808227/38133144-e32edfb8-3405-11e8-9118-37def5a78989.png">

The `"basic"` test differs from the two other tests in the `cli.test.js` (`--version` and `--help`) as it enters into the promise:

https://github.com/stylelint/stylelint/blob/67b062a24e8f0fe52058ffc8e07566d143a9c7e3/lib/cli.js#L448-L476

_Additionally, it enters into `getStdin()`_. Perhaps we need to account for these two things in the `"basic"` test.

@gucong3000 Do you have any ideas on how to address these issues in the `"basic"` test?

**Comment by ntwb** [comment]
Is https://github.com/stylelint/stylelint/commit/f60ab68d2efd6c9a58078aecc85a56d9bd9894af necessary? These tests have not been timing out on Travis CI following any of the other commits made direct to `master`?

**Comment by jeddy3** [comment]
> Is `f60ab68` necessary?

It was a workaround to get the release out.  Having fixed `cli.test.js` (it looks like [mocking `getStdin`](https://github.com/stylelint/stylelint/commit/55ab24d927c68306aea5a37e0e77b440567b19ed) did the trick), it then surfaced that both `001.test.js` and `005.test.js` were also timing out _during the release_. It's tricky to debug (hence the commits to `master`) as these three timeouts (`cli.test.js`, `001.test.js` and `005.test.js`) only happen during the release process. 

It could be the interplay between our tests and `npmpub`, and we should investigate further now that `9.2` is out. 

Increasing the timeout stopped the two tests from failing, but is stop gap solution.

<img width="482" alt="screen shot 2018-04-01 at 11 30 42" src="https://user-images.githubusercontent.com/808227/38172343-91a3b116-35a2-11e8-87e2-037fcbe2ef5e.png">

**Comment by jeddy3** [comment]
Done 🎉 (Sorry for the delay! We had planned to do the release last weekend, but, as you can see from the conversation above, there were a few bumps in the road).

External: @AndyOGo and @Bilie, thanks for your contributions to this release.

Core: @gucong3000 Thanks for the many fixes in this release! 

---

I'll likely make a couple of follow up issues for the things uncovered in this issue.
